### PR TITLE
#344: add generic list support to property

### LINKED ISSUE #326: Uninstall multiple tools at the same time
Enhance the UninstallCommandlet implementation to support uninstalling multiple tools at the same time.

### LINKED ISSUE #344: Add generic list support on Property
Currently every `Property` is designed to take exactly one value.
As a special feature, we have implemented `StringListProperty` that can take any number of `String` values.
With PR #258 we have now the need to also support list of other property types.
The idea is that instead of adding support for `ToolListProperty` and maybe many other `*ListProperty` classes on the long run causing quite some duplication and redundancy, it would be better to extend `Property` class itself with the ability to take multiple values.

The idea is that instead of 
https://github.com/devonfw/IDEasy/blob/b2754e933b89db43d9a661f451691b469c12672b/cli/src/main/java/com/devonfw/tools/ide/property/Property.java#L43
We will maintain a potential list of values:
```java
protected List<V> valueList;
```
The implementer may decide if it is easier to additionally have this `valueList` or to change `value` from `V` to `List<V>`.

To avoid breaking existing code the methods `getValue()` and `setValue(V)` should keep its API compatible.
So e.g. if we really have a `List` of values then `getValue()` returns the first value from the list if available or otherwise `null`.
Methods will be added like `getValue(int i)` and `getValueCount()` to access the values from within the `List`.

At construction time every property should have the capability to be configured to hold a single value and the existing constructors should stay compatible providing this behavior while optional additional arguments can set `multiValued` flag to `true` (and can be added later if ever needed potentially cardinalities like `minValueCount` and `maxValueCount` so the CLI could e.g. take only a number of values within this range).

The PR implementing this new feature should also remove `StringListProperty` and refactor code using it to use `StringProperty` with `multiValued` flag instead.
To keep CLI untouched the old behavior of calling `setValueAsString` multiple times actually adding values to `valueList` should be kept. The implementer may decide if `setValue` will also behave the same way. However, if `multiValued` is `false` the implementation must not add multiple values into the underlying list if `setValue[AsString]` is called multiple times.

### PR DESCRIPTION
**Pull request by ndemirca**
Closes #344 
Closes #326

### CODE REVIEW

**Review by hohwille — APPROVED** [review]
@ndemirca thanks for your PR. You have solved all the complex changes. Well done 👍 
I left quite some review comments for improvement.
You could simplify and reduce changes in existing code following our [coding-convetions](https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.adoc#refactorings).
I tried to make constructive changes and keep it still simple so even though it might look like a lot of comments, it should be possible to resolve quickly so we can merge.

File: cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java [hunk 1]
```diff
@@ -18,7 +18,20 @@ public class StringProperty extends Property<String> {
...
+    this(name, required, alias, false, null);
+  }
+
+  /**
+   * The constructor.
+   *
+   * @param name the {@link #getName() property name}.
+   * @param required the {@link #isRequired() required flag}.
+   * @param alias the {@link #getAlias() property alias}.
+   * @param multivalued
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java:30]:
```suggestion
   * @param multivalued the boolean flag about multiple arguments.
```

File: cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java [hunk 2]
```diff
@@ -28,10 +41,11 @@ public StringProperty(String name, boolean required, String alias) {
    * @param required the {@link #isRequired() required flag}.
    * @param alias the {@link #getAlias() property alias}.
    * @param validator the {@link Consumer} used to {@link #validate() validate} the {@link #getValue() value}.
+   * @param multivalued
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java:44]:
```suggestion
   * @param multivalued the boolean flag about multiple arguments
```

File: cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java [hunk 3]
```diff
@@ -46,4 +60,9 @@ public String parse(String valueAsString, IdeContext context) {
     return valueAsString;
   }
 
+  public String[] asArray() {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/property/StringProperty.java:63]:
```suggestion
  @Override
  public String[] asArray() {
```

**Review by hohwille — APPROVED** [review]
@ndemirca Great rework 👍 Thanks!
Are you still working on the constructors without the `boolean multiValued` parameter to keep signature compatibility and reduce the changes in commandlets?
Then I can wait a moment. Apart from that ready for merge.

**Comment by ndemirca** [comment]
@hohwille the changes about duplications in constructors is in the last commit. You can merge it, thanks!

**Review by jan-vcapgemini — APPROVED** [review]

**Pull request merged**